### ISSUE TITLE
Use a constant URI scheme

### ISSUE THREAD

**Issue by ndw** [cmarchand/maven-catalogBuilder-plugin#4]
I think the catalog-based approach that you're using is great, but I think using the artifact ID directly as the URI scheme is risky. It creates an unbounded number of different schemes, so it's impossible to tell from the URI what it is.

I think it would be better to have a fixed URI scheme (for example "mavenid") and encode the Maven information in the rest of the URI. For example,

````
  mavenid://artifactid/path/to/thing.xsl
````

I think you could even extend this to support groups and versions:

````
  mavenid://groupid:artifactid:version/path/to/thing.xsl
````

I don't remember the exact rules for the "authority" part of the URI, so a quick look at RFC 3986 is probably in order.

You could even pursue getting the 'mavenid' scheme (or whatever you call it) registered as an official scheme.

**Comment by cmarchand** [comment]
Thanks for your proposal.
RFC says :
````
authority   = [ userinfo "@" ] host [ ":" port ]
````
`host` is defined as 
````
IP-literal / IPv4address / reg-name
````
`groupId`**should** be based on domain-name, as a java package name, so it should be mapped to `reg-name`.
`port` is defined as
`````
port        = *DIGIT
`````
Definitively, `port` can not be used to store `artifactId`.

So, we need to encode all artifact information (groupId, artifactId, version) in `host` part.

Finally, reg-name is defined also in RFC3986. As :
````
reg-name    = *( unreserved / pct-encoded / sub-delims )
unreserved    = ALPHA / DIGIT / "-" / "." / "_" / "~"
sub-delims    = "!" / "$" / "&" / "'" / "(" / ")" / "*" / "+" / "," / ";" / "="
pct-encoded = "%" HEXDIG HEXDIG
````

In maven world, `groupId` are in fact java package names, and in repositories, they are organized as a directory tree structure : see [maven central](http://repo1.maven.org/maven2/). characters used in groupId conform to `sub-delims`, using only `.` as delimiter.

`artifactId` has no constraint in Maven pom schema, but in repositories, it is stored as a directory. So, there, common character exclusion applies, and there should not have `@`, `$`, and so on chars.

`release` are also stored in a directory, so same applies.

I think we could have a `authority` as this :
````
(groupId ";")? artifactId ( "(" version ")" )?
````

So a complete `<xsl:import/>` URI as :
````
    <xsl:import href="mavenid://eu.els.sie.lib.xsl.common;xslLib(3.1.5)/eflCommon.xsl"/>
````

Or a typical `<xsl:import/>` URI as :
````
    <xsl:import href="mavenid://eu.els.sie.lib.xsl.common;xslLib/eflCommon.xsl"/>
````
I'm not kind of using version in uris, that's why I put `version` between parenthesis, trying to discourage their use...

@ndw , @mricaud : your feelings ?

**Comment by mricaud** [comment]
Having both groupId and the artefact ID in the uri is definitively a good idea, if this is widely used there should not be any ambiguity.
Using a constant uri scheme is also quite a better alternativ than the original one, for sure.

Asking for my opinion always expose you to a lot of disgression from me ;)
I always need to check everything, something just ideas or thought. So this is a bit long and sometimes out of topic maybe, but please take the best and forget the rest! 

**About the uri scheme:**
- One could use the shorter "mvn" instead of maven (like the command line) ? `mvnid://`
- Or something to do with "dependency" ?  `mvndependency://` ?
- `mavenid://` sounds also good actualy, it is a "maven identification" uri 
- But I actualy like better the first proposal `artefact://`, just a feeling actually ... maybe more generic ? 
Should this kind of uri be also used with other dependency tools like gradle ?
- What about sticking with the well known `http://` ? This has not to be resolved directly:
    
    - `http://` is already used with xml namespace (it has nothing to do with http but it's an identifier uri too) 
    - and with publicID of a DTD for example. 

    But well, if we use `http://`  we should then probably use something else in the host to distinguish "real" http uris: starting with `http://artefact` ? So maybe not a good idea. 

    The good thing is that one may be able to publish artefact on the web just like @fgeorges does (but it's no more maven way of doing? both is possible ?) or just documentation. This would gave the alternative for people that don't use maven to call published artefact as well... That's a big deal but I think this is not the point of this project.

**About the groupId**
Sticking with the original groupId is definitively the best thing to do. 

On the other hand, one could also use the path form of the groupID:: eu/els/sie/xsl
(like the maven ".m2" directory does)
But that would make not much sens here I think.

**About the delimiter**
Just a questionn of taste: my preferences goes to "+" or "$"
One could also use / and another delimiter after // the artefactId: 
`eu.els.sie/xsl//` ?

**About the version**
Let me understand ...
I guess Maven is able to deal with 2 versions of the same groupId/artefactId, doesn't it?
But then a java programm is not able to load to different classes, as far as I understand ?
If Maven is ok with that, XML programms would also deal it with the help of the catalog.

That mean a pure xml project could use maven and multiple version dependency, as long as it's not mixed with java code (thinking about mixed artefact dependencies to classical java lib and new maven xml lib) ? Or multiple dependency would be reserved to xml project?

From a xml point of vue, I can quite easily imagine using 2 differents versions of a xml dependency in a big programm. That would actually be really usefull: one would then be able to switch independently each xslt (or xsd, etc.) to a special version. That's more safe.
We face this problem today at ELS: switching to a new version of a xsl lib oblige us to look at every piece of code. Which is quite a constraint, especialy when we use a maven parent pom to hold the dependency for every maven submodule.
One more argument is that deep dependencies (dependency of dependency of dependency) make sometimes difficult to only have a single version for a common lib. Maven can deal with that making choice for us, but it might lead to strange behaviours (like saxon says with multiple import of the same xsl;))?

It's a question of granularity: if we can add the version to the URI, every developper of a common project may use the appropriate version of the lib without disturbing other's work.

So I'm definitely in favour of _giving the possibility_ to add the version within the uri.

**Finaly**
After all this (long) considerations, if I am to make a choice I would use:
`<xsl:import href="artefactId://eu.els.sie.lib.xsl.common+xslLib$3.1.5/eflCommon.xsl"/>`

**Comment by cmarchand** [comment]
Have a look at [RFC 3986](https://tools.ietf.org/html/rfc3986#section-3), uri is defined as this :
````
      URI         = scheme ":" hier-part [ "?" query ] [ "#" fragment ]

      hier-part   = "//" authority path-abempty
                     / path-absolute
                     / path-rootless
                     / path-empty
````
So `/` can not be used as a delimiter. `+` & `$` sounds good.

Now, letting user defining its own URI pattern is not good. I propose to have only three URI forms :
 - **full** : `dependency://groupId+artifactId$version/path/to/resource`
 - **standard** : `dependency://groupId+artifactId/path/to/resource`
 - **compact** : `dependency://artifactId/path/to/resource`

The default URI form being **standard**.

By default, plugin will only generate **standard** URI form catalog entries, but it could alse generate the two other forms, at once :

````
    <configuration>
      <uriPatterns>
        <uriPattern>full</uriPattern>
        <uriPattern>standard</uriPattern>
        <uriPattern>compact</uriPattern>
      </uriPatterns>
    </configuration>
````

Finally, a catalog has a lot of rewrite systems, so we could specify which generate :
````
    <configuration>
      <generates>
        <generate>public</generate>
        <generate>system</generate>
        <generate>rewriteSystem</generate>
        <generate>delegatePublic</generate>
        <generate>delegateSystem</generate>
        <generate>uri</generate>
        <generate>rewriteURI</generate>
        <generate>delegateURI</generate>
      </generates>
    </configuration>
````
If `generates` is not specified, it uses `rewriteSystem` and `rewriteURI`.

Using `version` in code is definitely a bad practise. Doing so, you are required to modify your code when you change a dependency version, and that is what maven troes to avoid. Java is not able, today, to have to different releases of a same jar in classpath, or this requires complex systems, like OSGI.  
We could define the URI scheme to allow version in, for a future use, with [aether](https://wiki.eclipse.org/Aether/What_Is_Aether) for example, but in ELS use of dependency management, this should be strictly avoided, that's why I'd like to define `standard` URI form as default.

Thoughts ?

**Comment by ndw** [comment]
Well, since @mricaud proposed it, I'm going to second the idea. You could just use http(s):

     https://[yourhostnamehere]/groupId/artifactId/path/to/resource?version=version

I'm a little unclear on why there's ever any advantage in having the `artifactId` without its corresponding `groupId`. (And I don't feel strongly about making version a parameter, it's just an idea.)

An advantage of an https URI is that it can resolve and return useful information if someone clicks on it or types it into a browser.

I'd happily setup mvn.xmlresolver.org if you'd like.

**Comment by cmarchand** [comment]
We can't use `http(s)` for this, as there is no dedicated server to locate a dependency. From a client to another, you may have different settings, and those settings are used by maven to locate the dependency artifact.

So, I preferer use `dependency://`, which denotes exactly a dependency reference, but delegates dependency resolution to implementation ; I do use Maven, but one could use anything else.

Release 2.0.0-RC1 released.
Type `mvn top.marchand.xml.maven.catalogBuilder-maven-plugin:help` to see all configuration parameters. I'm gonna to update wiki ASAP.

**Comment by cmarchand** [comment]
[Wiki](https://github.com/cmarchand/maven-catalogBuilder-plugin/wiki) has been updated.

**Comment by ndw** [comment]
There is nothing about the http(s) scheme that requires a server or runtime access. Nevertheless, it is often remarkably convenient that such access is possible by humans with browsers.

The name of this repo is `maven-...` so I think my assumption that was about Maven was justified :-), but if you want a more generic hostname, that's fine too. `packages.xmlresolver.org`?

Looking at the wiki, how confident are you that "+" and "$" can *never* appear in `GroupId` or `ArtifactId` values?

What's the motivation for the shortcut form that uses only artifactId? Given that these are generated, isn't it simpler and safer to always generate the long form? It seems to me that the short form is just a problem waiting to happen. Consider the first time someone wants to use 'someThing' artifact from two different groups?

**Comment by cmarchand** [comment]
@ndw :
I'm not confident that "+" and "$" will *never* appear. But, `groupId` **must** conform to a java package name, so I'm confident on groupId part.
On artifact part, when I search [The Central Repository](http://search.maven.org) for artifact that contains a "+" or a "$", nothing's found.

So... I'm quite confident...

Could you be more precise with your idea on `packages.xmlresolver.org` ?

**Comment by cmarchand** [comment]
I think you are right, and I'm gonna remove the `compact` form in RC2

**Comment by ndw** [comment]
My idea for packages.xmlresolver.org is that you can use URIs of this form:

    https://packages.xmlresolver.org/[what the spec says]

Those will work perfectly fine with the catalog resolver, no network access necessary.

But if someone wonders, "what is this thing?" and points a web browser at it, we can arrange for the site to return useful documentation about what it is, relevant pointers, etc.

I guess I'm volunteering to do that part. :-)

**Comment by cmarchand** [comment]
Solved in release > 2.00.00
